src/main/kotlin/external_app.kt

```diff
@@ -0,0 +1,122 @@
+import app.filesystem.AudioStore
+import app.filesystem.DirectoryProvider
+import java.io.File
+import java.io.Reader
+import java.io.InputStreamReader
+import java.io.BufferedReader
+import java.util.*
+
+
+fun main(args:Array<String>) {
+
+    val os = System.getProperty("os.name").toUpperCase() //current user's OS
+
+    val locale = Locale.getDefault()
+
+    val labels= ResourceBundle.getBundle("Resources", locale) //accesses the properties file for the current project
+
+    val program = labels.getString("Ext_Recording_Application_Name") //gets the external recording application name
+
+    val audioStore = AudioStore()
+
+    //create the take recording file (tests AudioStore class functions)
+    val take_recording_file = audioStore.createTakeFile("en", "ot", "gen", "01", "01", "02")
+
+    if(!audioStore.checkTakeLocationAndName(take_recording_file)){
+        println("Given file info is not valid")
+        return
+    }
+
+    //open the specified external application record the take file
+    try {
+        println("Opening audacity")
+
+        //open Audacity (must be installed on the user's computer)
+        try{
+            //determines and executes the appropriate command to open Audacity
+            if (os.contains("WIN")){
+                //need to fix: Windows does not recognize start command for external apps
+                val process = ProcessBuilder("start", program, take_recording_file.absolutePath)
+                process.start()
+            }
+            else if (os.contains("MAC")){
+                val process = ProcessBuilder("open", "-a", program, take_recording_file.absolutePath)
+                process.start()
+            }
+            else { //Linux OS
+                val process = ProcessBuilder(program, take_recording_file.absolutePath)
+                process.start()
+            }
+            //calls helper function to get the current status of Audacity
+            var status = getCurrStatus(os, true)
+            while(status == 1){
+                status = getCurrStatus(os, false)
```

Comment:
why not just use waitFor()? It should block the thread until the process returns. We could even have a dialog up that says "Please finish work in audacity before continuing" or something before you lock the UI.

Reply:
I tried to use waitFor() on the command to open Audacity, but it kept on returning 0 right away and I assumed it was because it had already executed the command and audacity had already opened. I couldn't figure out how to get the exit code of audacity itself, so i just wrote getCurrStatus to go through the currently running processes and see if audacity was part of them or not to determine its status.

Reply:
I don't know, I'd need to play around with it I guess. Your solution is definitely creative, I just don't know how fault tolerant it is (and maybe it is). One thing you probably would prefer to do for performance reasons is to not do a totally busy wait (it's going to be CONSTANTLY checking the status, and that status won't change frequently; you probably want to do a slight Thread.sleep for like, a second or a little less in between checks)